rcl/src/rcl/graph.c

```diff
@@ -452,6 +456,158 @@ rcl_count_subscribers(
   return rcl_convert_rmw_ret_to_rcl_ret(rmw_ret);
 }
 
+typedef rcl_ret_t (* count_entities_func_t)(
+  const rcl_node_t * node,
+  const char * topic_name,
+  size_t * count);
+
+rcl_ret_t
+_rcl_wait_for_entities(
+  const rcl_node_t * node,
+  rcl_allocator_t * allocator,
+  const char * topic_name,
+  const size_t expected_count,
+  rcutils_duration_value_t timeout,
+  bool * success,
+  count_entities_func_t count_entities_func)
+{
+  if (!rcl_node_is_valid(node)) {
+    return RCL_RET_NODE_INVALID;
+  }
+  RCL_CHECK_ALLOCATOR_WITH_MSG(allocator, "invalid allocator", return RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(topic_name, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(success, RCL_RET_INVALID_ARGUMENT);
+
+  *success = false;
+
+  // We can avoid waiting if there are already the expected number of publishers
+  size_t count = 0u;
+  rcl_ret_t count_ret = count_entities_func(node, topic_name, &count);
+  if (count_ret != RCL_RET_OK) {
+    // Error message already set
+    return count_ret;
+  }
+  if (expected_count <= count) {
+    *success = true;
+    return RCL_RET_OK;
+  }
+
+  // Create a wait set and add the node graph guard condition to it
+  rcl_wait_set_t wait_set = rcl_get_zero_initialized_wait_set();
+  rcl_ret_t ret = rcl_wait_set_init(
+    &wait_set, 0, 1, 0, 0, 0, 0, node->context, *allocator);
+  if (ret != RCL_RET_OK) {
+    // Error message already set
+    return ret;
+  }
+
+  const rcl_guard_condition_t * guard_condition = rcl_node_get_graph_guard_condition(node);
+  if (!guard_condition) {
+    // Error message already set
+    return RCL_RET_ERROR;
+  }
+
+  // Add it to the wait set
+  ret = rcl_wait_set_add_guard_condition(&wait_set, guard_condition, NULL);
+  if (ret != RCL_RET_OK) {
+    // Error message already set
+    return ret;
+  }
+
+  // Get current time
+  rcutils_time_point_value_t start;
+  rcutils_ret_t time_ret = rcutils_steady_time_now(&start);
```

Comment:
I think a kind of meta issue we have to resolve in rcl (not here, and it has come up elsewhere), is that we don't know really what kind of clock the wait set uses to figure its timeout. I think it says system atm, and some do that and other don't? I don't remember off the top of my head and don't have time to dig atm. Anyway I mention that because here we're using steady time and the wait might be using system time, which may give some odd results.

But since this isn't an issue unless you're majorly shifting the date while running this function, it seems low risk atm.

Reply:
I can't find docs in rcl about the clock, but rmw is documented to use system time:

https://github.com/ros2/rmw/blob/ace271650ad39a656d81239050c9639c4190f077/rmw/include/rmw/rmw.h#L2371-L2374

Maybe it's better to use system time here instead; then we're consistent with `rcl_wait` (assuming the RMW is respecting the docs).

Reply:
Yeah may be safer.

Somewhere I expressed the desire that rmw_wait would use steady time, but ultimately we're at the mercy of the underlying middleware.

Reply:
Switched to system time: af14fc4